src/AggregateFunctions/AggregateFunctionMap.h

```diff
@@ -0,0 +1,207 @@
+            nested_func->add(nested_place, nested_columns, offset + i, arena);
+        }
+    }
+
+    void merge(AggregateDataPtr place, ConstAggregateDataPtr rhs, Arena * arena) const override
+    {
+        auto & merged_maps = this->data(place).merged_maps;
+        const auto & rhs_maps = this->data(rhs).merged_maps;
+
+        for (const auto & elem : rhs_maps)
+        {
+            const auto & it = merged_maps.find(elem.first);
+
+            if (it != merged_maps.end())
+            {
+                nested_func->merge(it->second, elem.second, arena);
+            }
+            else
+                merged_maps[elem.first] = elem.second;
+        }
+    }
+
+    void serialize(ConstAggregateDataPtr place, WriteBuffer & buf) const override
+    {
+        auto & merged_maps = this->data(place).merged_maps;
+        writeVarUInt(merged_maps.size(), buf);
+
+        for (const auto & elem : merged_maps)
+        {
+            this->data(place).writeKey(elem.first, buf);
+            nested_func->serialize(elem.second, buf);
+        }
+    }
+
+    void deserialize(AggregateDataPtr place, ReadBuffer & buf, Arena * arena) const override
+    {
+        auto & merged_maps = this->data(place).merged_maps;
+        size_t size;
+
+        readVarUInt(size, buf);
+        for (size_t i = 0; i < size; ++i)
+        {
+            KeyType key;
+            AggregateDataPtr nested_place;
+
+            this->data(place).readKey(key, buf);
+            nested_place = arena->alloc(nested_func->sizeOfData());
+            nested_func->create(nested_place);
+            merged_maps.emplace(key, nested_place);
+            nested_func->deserialize(nested_place, buf, arena);
+        }
+    }
+
+    void insertResultInto(AggregateDataPtr place, IColumn & to, Arena * arena) const override
+    {
+        auto & map_column = assert_cast<ColumnMap &>(to);
+        auto & nested_column = map_column.getNestedColumn();
+        auto & nested_data_column = map_column.getNestedData();
+
+        auto & key_column = nested_data_column.getColumn(0);
+        auto & val_column = nested_data_column.getColumn(1);
+
+        auto & merged_maps = this->data(place).merged_maps;
+
+        size_t res_offset = 0;
+
+        // sort the keys
+        std::vector<KeyType> keys;
+        keys.reserve(merged_maps.size());
+        for (auto & it : merged_maps)
+        {
+            keys.push_back(it.first);
+        }
+        std::sort(keys.begin(), keys.end());
+
+        // insert using sorted keys to result column
+        for (auto & key : keys)
+        {
+            res_offset++;
+            key_column.insert(key);
+            nested_func->insertResultInto(merged_maps[key], val_column, arena);
+        }
+
+        IColumn::Offsets & res_offsets = nested_column.getOffsets();
+        auto last_offset = res_offsets[res_offsets.size() - 1];
+        res_offsets.push_back(last_offset + res_offset);
```

Comment:
Probably we can do just `res_offsets.push_back(val_column.size());`

Reply:
Yeah, it works, done.